src/dht/mod.rs

```diff
@@ -54,28 +54,66 @@ impl Dht {
         let id = self.id.clone();
 
         let bootstrap_futures = addrs.into_iter().map(move |addr| {
-            send_transport
-                .ping(id.clone(), addr.clone().into())
-                .and_then(move |id| Ok(Node::new(id, addr.clone().into())))
+            Self::discover_node(
+                addr,
+                id.clone(),
+                send_transport.clone(),
+                routing_table_arc.clone(),
+            )
         });
 
-        let bootstrap_future =
-            future::join_all(bootstrap_futures).and_then(move |nodes| -> Result<()> {
-                let mut routing_table = routing_table_arc.lock()?;
-
-                nodes
-                    .into_iter()
-                    .for_each(|node| routing_table.add_node(node));
-
-                Ok(())
-            });
+        let bootstrap_future = future::join_all(bootstrap_futures).and_then(|_| Ok(()));
 
         bootstrap_future
 
         // TODO:
         // * Query Node for Self Until Some Amount of Nodes Have Been Successfully Added
     }
 
+    fn discover_node(
+        addr: SocketAddrV4,
+        self_id: NodeID,
+        send_transport: Arc<SendTransport>,
+        routing_table_arc: Arc<Mutex<RoutingTable>>,
+    ) -> Box<Future<Item = (), Error = Error> + Send> {
+        let cloned_routing_table = routing_table_arc.clone();
+
+        let fut = send_transport
+            .find_node(self_id.clone(), addr.clone().into(), self_id.clone())
+            .timeout(Duration::from_secs(5))
+            .then(|result| match result {
+                Ok(res) => Ok(Some(res)),
+                Err(..) => Ok(None),
```

Comment:
Hmm, errors aren't bubbled up from here. They are ignored. Should probably do something about this maybe?